### PR TITLE
fix(http):fix jsonp callbackname is undefined

### LINKED ISSUE #14267: Jsonp error `Uncaught ReferenceError: __ng_jsonp____req0_finished is not defined` after https://github.com/angular/angular/pull/13219
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

hello, after upgrade to 2.4.6, I get `Uncaught ReferenceError: __ng_jsonp____req0_finished is not defined` as https://github.com/angular/angular/pull/13219#discussion_r90797936
image

**Plunkr**
http://plnkr.co/edit/WcRnPrJHa04rJy3xTIky?p=preview

And my code is
```
search(key: string) {
  const params = new URLSearchParams()
  params.set('key', key)
  params.set('jsonpCallback', 'JSONP_CALLBACK')
  return this.jsonp.get('https://c.y.qq.com/splcloud/fcgi-bin/smartbox_new.fcg?' + params)
  .toPromise()
  .then(res => {
    console.log(res)
    return []
  })
}
```

### PR DESCRIPTION
**Pull request by doxiaodong**
fix https://github.com/angular/angular/issues/14267

### CODE REVIEW

**Comment by DzmitryShylovich** [comment]
need to add a test to ensure that callback is actually executed to prevent such situations in the future

**Comment by doxiaodong** [comment]
I update the callback name test, it can be any function name.

File: modules/@angular/http/src/backends/browser_jsonp.ts [hunk 1]
```diff
@@ -32,17 +30,13 @@ export class BrowserJsonp {
 
   nextRequestID(): string { return `__req${_nextRequestId++}`; }
 
-  requestCallback(id: string): string { return `${JSONP_HOME}${id}_finished`; }
+  requestCallback(id: string): string { return _getJsonpCallbackName(id); }
```
Comment (DzmitryShylovich) [inline modules/@angular/http/src/backends/browser_jsonp.ts]: revert

File: modules/@angular/http/test/backends/jsonp_backend_spec.ts [hunk 2]
```diff
@@ -70,10 +70,11 @@ export function main() {
       expect(instance).toBeAnInstanceOf(JSONPConnection);
     });
 
-    it('callback name should not contain dots', () => {
+    it('callback name should be a function name', () => {
```
Comment (DzmitryShylovich) [inline modules/@angular/http/test/backends/jsonp_backend_spec.ts]: as we saw testing callback name doesn't actually verify anything.
Need a test for `Uncaught ReferenceError: __ng_jsonp____req0_finished is not defined` arror

Comment (doxiaodong) [inline modules/@angular/http/test/backends/jsonp_backend_spec.ts]: this callback name can be any validate function name, and I think it's enough.  What to cause `__ng_jsonp____req0_finished is not defined` is the error of the data https://github.com/angular/angular/blob/master/modules/playground/src/jsonp/people.json#L2

**Review by alxhub — APPROVED** [review]

**Comment by DzmitryShylovich** [comment]
@alxhub this pr will add callbacks to the global namespace (window). It an be very unexpected :)
@mhevery said he will revert https://github.com/angular/angular/commit/9e5617e41e222d29c3952ba71a401804030cd153 (commit that introduced the bug)
and it's already fixed in 4.0 so I don't think you should merge this

**Comment by alxhub** [comment]
I actually think it's okay, and the behavior is correct.

 The callbacks are only there while the request is in progress, and some APIs have issues with dots in the callback names.

**Comment by DzmitryShylovich** [comment]
ok :)

**Comment by doxiaodong** [comment]
so, If I need to resolve the conflicts?

**Comment by doxiaodong** [comment]
how about this?

**Comment by DzmitryShylovich** [comment]
@doxiaodong it will be fixed in the new http module in 4.0
so I think you can close this

**Comment by splincode** [comment]
@doxiaodong What is the status of the task? `@angular/http is deprecated`, please closed PR

**Comment by doxiaodong** [comment]
ok

**Pull request closed without merging**